components/navigable-container/README.md

```diff
@@ -0,0 +1,156 @@
+NavigableContainer
+=============
+
+`NavigableContainer` is a React component to render a container navigable using the keyboard. Only things that are focusable can be navigated to. It will currently always be a `div`.
+
+`NavigableContainer` is exported as three convenience classes: `NavigableMenu`, `NavigableGrid` and `TabbableContainer`. `NavigableContainer` itself is not exported. `NavigableMenu`, `NavigableGrid`, and `TabbableContainer` have some shared props, and some individual props. Any other props will be passed through to the `div`.
+
+----
+
+## Shared Props
+
+These are the props that `NavigableMenu`, `NavigableGrid` and `TabbableContainer` all share.
+
+
+### initialSelector
+
+A selector which tells the component what to focus first inside it. If the selector is not matched, the first focusable item will be focused. This can be useful to focus the previously selected descendant when this component get focused.
+
+- Type: `String`
+- Required: No
+
+### onNavigate
+
+A callback invoked when the menu navigates to one of its children passing the index and child as an argument
+
+- Type: `Function`
+- Required: No
+
+### deep
+
+A boolean to look for navigable children in the direct children or any descendant. True means that any descendant can be considered navigable, and false means only direct children are considered.
+
+- Type: `Boolean`
+- Required: No
+- default: false
+
+
+### widget
+
+A boolean which specifies whether or not the focusable elements in this component might contain the focus, rather than have the focus. This is often used for `TabbableContainer` layouts that might have arrow key navigation inside each tabstop.
```

Comment:
Not sure I understand, could you clarify a bit?

Reply:
Yep, I'll clarify here, and then it makes sense here, I'll (try to remember to) update the readme.

The way this component is setup is designed to simplify *composing* different navigation models. Because it's the example in the other PR, it's easiest to to use the InserterMenu as an example.

The InserterMenu has a tabstop at the row of tabs, and a tabstop at the grid for each category, and a tabstop for the searcher. Now, the outer container of the inserter menu is just a `TabbableContainer`, with only knowledge of the three tabstops. However, once the row of tabs gets focused (for example), its own navigation activates. The row of tabs is a NavigableMenu, so it focused the selected tab inside.

Now, if the user presses tab, the NavigableMenu doesn't handle that, so it safely bubbles to the TabbableContainer. Now, the tabbablecontainer has to work out which of its tabstops currently has focus. But none of its tabstops currently have focus. A button which is *inside* one of its current tabstops has focus, but the tabstop itself doesn't. That's because the tabstop for the row of tabs is a widget.

It's part of the purpose of these modes. I've written something very similar to this before in a UI library I've made previously, so I'm probably guilty of trying to replicate something I feel works fairly well. Navigation often needs to be composable. Something on the outside tabs in between various tabtops, and those tabstops themselves have arrow key navigation (is the common example). Some things are even more complex. By allowing for different modes, you make it possible to compose different types of navigation within each other, and allow bubbling of unhandled keys (by a particular navigation mode) pass the keys to the outer navigation layer which does care about those keys. Once you set it up this way, all types of navigation become fairly simple to implement. If you discover that you need to handle ENTER or ESCAPE or something similar, or different types of keys for switching focus (HOME and END), then you just have to either add modes, or add configuration to a mode.

I realised I just tried to explain what modes are for. Does any of that make sense?